### ISSUE TITLE
Question about computing `Ratio`

### ISSUE THREAD

**Issue by RuizhouLiu** [XueZeyue/DanceGRPO#114]
Great work! When I run wan 2.1, the way of computing `ratio` confuses me. The code makes $\pi(a_t | s_t) / \pi_{old}(a_t | s_t)$ always be `1.`.

```
    train_timesteps = int(len(samples["timesteps"][0])*args.timestep_fraction)
    for i,sample in list(enumerate(samples_batched_list)):
        for _ in range(train_timesteps):
            clip_range = args.clip_range
            adv_clip_max = args.adv_clip_max
            new_log_probs = grpo_one_step(
                args,
                sample["latents"][:,_],
                sample["next_latents"][:,_],
                sample["encoder_hidden_states"],
                sample["negative_prompt_embeds"],
                transformer,
                sample["timesteps"][:,_],
                perms[i][_],
                sigma_schedule,
            )

            advantages = torch.clamp(
                sample["advantages"],
                -adv_clip_max,
                adv_clip_max,
            )

            ratio = torch.exp(new_log_probs - sample["log_probs"][:,_])

            unclipped_loss = -advantages * ratio
            clipped_loss = -advantages * torch.clamp(
                ratio,
                1.0 - clip_range,
                1.0 + clip_range,
            )
            loss = torch.mean(torch.maximum(unclipped_loss, clipped_loss)) / (args.gradient_accumulation_steps * train_timesteps)

            loss.backward()
            avg_loss = loss.detach().clone()
            dist.all_reduce(avg_loss, op=dist.ReduceOp.AVG)
            total_loss += avg_loss.item()
        if (i+1)%args.gradient_accumulation_steps==0:
            grad_norm = torch.nn.utils.clip_grad_norm_(transformer.parameters(), max_norm=max_grad_norm)
            optimizer.step()
            lr_scheduler.step()
            optimizer.zero_grad()
```

I look for same code in FlowGRPO, [see code ](https://github.com/yifan123/flow_grpo/blob/9745241cf1101407e9b518dd4f10498092150522/scripts/train_flux.py#L895.), they update params after loss.backward() immediately,

**Comment by Jayce-Ping** [comment]
Actually, the code you linked:

https://github.com/yifan123/flow_grpo/blob/9745241cf1101407e9b518dd4f10498092150522/scripts/train_flux.py#L833-L901


is wrapped in the context `with accelerator.accumulate(transformer):` such that the `gradient_accumulation_steps` works. Actually, Flow-GRPO update parameters two times (two steps) per epoch.

`ratio` can always be `1.00` if the training is always on-policy. It is normal if the reward increases.
